Stable Diffusion XL on Mac with Advanced Core ML Quantization (#1350)
huggingface/blog: stable-diffusion-xl-coreml.md

---
title: "Stable Diffusion XL on Mac with Advanced Core ML Quantization"
thumbnail: /blog/assets/stable-diffusion-xl-coreml/thumbnail.png
authors:
- user: pcuenq
- user: Atila
  guest: true
---

# Stable Diffusion XL on Mac with Advanced Core ML Quantization

<!-- {blog_metadata} -->
<!-- {authors} -->

> Stable Diffusion XL was released yesterday and it’s awesome. It can generate large (1024x1024) high quality images, adherence to the prompt has been improved with some new tricks, it can effortlessly produce very dark or very bright images thanks to the latest research on noise schedulers, and it’s open source!

Reviewer: ```suggestion
Stable Diffusion XL was released yesterday and it’s awesome. It can generate large (1024x1024) high-quality images; adherence to the prompt has been improved with some new tricks; it can effortlessly produce very dark or very bright images thanks to the latest research on noise schedulers; and it’s open source!
```

I'm not a grammar expert, but when using commas it was a little difficult to read. It felt like it said:
"It can:
- generate images
- adherence to ...
- it can ...
- and it's open source"

Semi-colons maybe help separate the ideas out a bit (kind of like listing without using a list)

> The downside is that the model is much bigger, and hence slower and harder to run on consumer hardware. Using the latest release of the Hugging Face diffusers library, you can run Stable Diffusion XL on CUDA hardware in 16 GB of GPU RAM, making it possible to use it on Colab’s free tier.

Reviewer: ```suggestion
The downside is that the model is much bigger, and therefore slower and more difficult to run on consumer hardware. Using the latest release of the Hugging Face diffusers library, you can run Stable Diffusion XL on CUDA hardware in 16 GB of GPU RAM, making it possible to use it on Colab’s free tier.
```

> The past few months have shown that people are very clearly interested in running ML models locally for a variety of reasons, including privacy, convenience, easier experimentation, or unmetered use. We’ve been working hard at both Apple and Hugging Face to explore this space. We’ve shown how to run Stable Diffusion on Apple Silicon, or how to leverage the latest advancements in Core ML to improve size and performance with 6-bit palettization.

Reviewer: ```suggestion
The past few months have shown that people are very clearly interested in running ML models locally for a variety of reasons, including privacy, convenience, easier experimentation, or unmetered use. We’ve been working hard at both Apple and Hugging Face to explore this space. We’ve shown how to run Stable Diffusion on Apple Silicon, and how to leverage the latest advancements in Core ML to improve size and performance with 6-bit palettization.
```

For Stable Diffusion XL we’ve done a few things:
* Ported the base model to Core ML so you can use it in your native Swift apps.

> Updated Apple’s conversion and inference repo so you can convert the models yourself, including any fine-tunes you’re interested in.

Reviewer: ```suggestion
* Updated Apple’s conversion and inference repo so you can convert the models yourself, including any fine-tunes you’re interested in.
```

- [How are Mixed-Bit Recipes Created?](#how-are-mixed-bit-recipes-created)
- [Converting Fine-Tuned Models](#converting-fine-tuned-models)
- [Resources](#resources)

## Using SD XL Models from the Hugging Face Hub

As part of this release, we published two different versions of Stable Diffusion XL in Core ML.
- [`apple/coreml-stable-diffusion-xl-base`](https://huggingface.co/apple/coreml-stable-diffusion-xl-base) is a complete pipeline, without any quantization.
- [`apple/coreml-stable-diffusion-mixed-bit-palettization`](https://huggingface.co/apple/coreml-stable-diffusion-mixed-bit-palettization) contains (among other artifacts) a complete pipeline where the UNet has been replaced with a mixed-bit palettization _recipe_ that achieves a compression equivalent to 4.5 bits per parameter. Size went down from 4.8 to 1.4 GB, a 71% reduction, and in our opinion quality is still great.

Either model can be tested using Apple’s [Swift command-line inference app](https://github.com/apple/ml-stable-diffusion#inference), or Hugging Face’s [demo app](https://github.com/huggingface/swift-coreml-diffusers). This is an example of the latter using the mixed-bit pipeline:

> [screenshot]

Reviewer: ^ I assume a TODO 😄

> As with previous Stable Diffusion releases, we expect the community to come up with novel fine-tuned versions for different domains, and many of them will be converted to Core ML. You can keep an eye on [this filter in the Hub]() to explore!

Reviewer: Missing link

Stable Diffusion XL works on Apple Silicon Macs running the public beta of macOS 14. It uses the `ORIGINAL` attention implementation, which is intended for CPU + GPU compute units. Note that the refiner stage has not been ported yet.

> For reference, these are the performance figures we got on a few computers:

Reviewer: ```suggestion
For reference, these are the performance figures we achieved on different devices:
```

devices/computers

|        Device         | `--compute-unit`| `--attention-implementation` | End-to-End Latency (s) | Diffusion Speed (iter/s) |
| --------------------- | --------------- | ---------------------------- | ---------------------- | ------------------------ |
| Mac Studio (M2 Ultra) | `CPU_AND_GPU`   |      `ORIGINAL`              |      47                |        1.11              |
| Mac Studio (M1 Ultra) | `CPU_AND_GPU`   |      `ORIGINAL`              |      58                |        0.89              |
| MacBook Pro (M1 Max)  | `CPU_AND_GPU`   |      `ORIGINAL`              |     113                |        0.46              |

## What is Mixed-Bit Palettization?

> Last month we discussed 6-bit palettization, a post-training quantization method that converts 16-bit weights to just 6-bit per parameter. This achieves an important reduction in model size, but going beyond that is tricky because model quality becomes more and more impacted as the number of bits is decreased.

Reviewer: maybe link to that article

Author: Yeah, it was a personal to-do, fixed while you were reviewing :)

Author: Actually I had missed this one, thanks!

Using this method, we can achieve effective quantizations of, for example, 2.8 bits on average, and we measure the impact on degradation for every combination we try. This allows us to be better informed about the best quantization to use for our target quality and size budgets.

To illustrate the method, let’s consider the following alternative quantization “recipes” (we’ll see later how they were generated):

```json
{
  "model_version": "stabilityai/stable-diffusion-xl-base-1.0",
  "baselines": {
    "original": 82.2,
    "linear_8bit": 66.025,
    "recipe_6.55_bit_mixedpalette": 79.9,
    "recipe_4.50_bit_mixedpalette": 75.8,
    "recipe_3.41_bit_mixedpalette": 71.7,
  },
}
```

> What this tells us is that the original model quality, as measured by PSNR in float16, is of about 82 dB. Performing a naïve 8-bit linear quantization drops it to 66 dB. But then we have a recipe that achieves 6.55 bits per parameter, on average, while keeping PSNR at 80 dB. The second and third recipes further reduce the model size, while still sustaining a PSNR larger than that from the 8-bit linear quantization.

Reviewer: ```suggestion
What this tells us is that the original model quality, as measured by PSNR in float16, is about 82 dB. Performing a naïve 8-bit linear quantization drops it to 66 dB. But then we have a recipe that achieves 6.55 bits per parameter, on average, while keeping PSNR at 80 dB. The second and third recipes further reduce the model size, while still sustaining a PSNR larger than that of the 8-bit linear quantization.
```

For visual examples, these are the results on prompt `a high quality photo of a surfing dog` running each one of the three recipes with the same seed:

| 3.41-bit | 4.50-bit | 6.55-bit | 16-bit (original) |
| :-------:| :-------:| :-------:| :----------------:|

> | ![3.41](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/stable-diffusion-xl-coreml/a_high_quality_photo_of_a_surfing_dog.7667.final_3.41-bits.jpg) | ![4.50](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/stable-diffusion-xl-coreml/a_high_quality_photo_of_a_surfing_dog.7667.final_4.50-bits.jpg) | ![6.55](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/stable-diffusion-xl-coreml/a_high_quality_photo_of_a_surfing_dog.7667.final_6.55-bits.jpg) | ![16-bit](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/stable-diffusion-xl-coreml/a_high_quality_photo_of_a_surfing_dog.7667.final_original.jpg) |

Reviewer: the last link seems broken:

![image](https://github.com/huggingface/blog/assets/26504141/9453b822-0190-4b11-ad5b-e935a7ecbdf4)

* [Recipes for Stable Diffusion 1.5](https://huggingface.co/apple/coreml-stable-diffusion-mixed-bit-palettization/blob/main/recipes/runwayml-stable-diffusion-v1-5_palettization_recipe.json)
* [Recipes for Stable Diffusion 2.1](https://huggingface.co/apple/coreml-stable-diffusion-mixed-bit-palettization/blob/main/recipes/stabilityai-stable-diffusion-2-1-base_palettization_recipe.json)
* [Recipes for Stable Diffusion XL 1.0 base](https://huggingface.co/apple/coreml-stable-diffusion-mixed-bit-palettization/blob/main/recipes/stabilityai-stable-diffusion-xl-base-1.0_palettization_recipe.json)

You can download and apply them locally to experiment.

In addition, we also applied the three best recipes from the Stable Diffusion XL analysis to the Core ML version of the UNet, and published them [here](https://huggingface.co/apple/coreml-stable-diffusion-mixed-bit-palettization/tree/main/unet-mbp-sdxl-1-base). Feel free to play with them and see how they work for you!

Finally, as mentioned in the introduction, we created a [complete Stable Diffusion XL Core ML pipeline](https://huggingface.co/apple/coreml-stable-diffusion-mixed-bit-palettization) that uses the `4.5` recipe.

### Resources

> 

Reviewer: another TODO
